### DISCUSSION TITLE
How to have a common working directory in code chunks and rendering in a r project?

### DISCUSSION THREAD

**Discussion by victorfeagins** [quarto-dev/quarto-cli#9287] [Q&A]
### Description

I often for class have a single R project and subfolders for each assignment. By default when running code chunks it uses the project as the working directory and when rendering it uses the folder where the file is saved. I usually use the `here` package to get around this. 

Is there a way without using the `here` package to succulently to set the working directory to use the project directory or the subfolder as working directory in code chunks and rendering?

**Comment by mcanouil** [comment]
The default is not project, see <https://quarto.org/docs/projects/code-execution.html#working-dir>.
You can set it to project, but it does not change anything about how R code are evaluated in an R process started by you.

> By default when running code chunks it uses the project as the working directory

when you run R code or any code interactively, it is independent of the context, as R process does not read any of the Quarto stuff.
It's not a default, it's R behaviour. The working directory is set to where R is started.

If you want your working directory to not be the project root or the file location, there is no other choice for you than to set the working directory yourself in a code cell using whatever means you feel is appropriate (here, withr, base R).

**Reply by victorfeagins** [reply]
Setting the working directory using in `_quarto.yml` works great. But it changes how the render button works in Rstudio as it tries to render the whole project with an updated individual file (I think). Do you know how to adjust the behavior of the render button?

**Reply by cderv** [reply]
@victorfeagins it seems this could be that the quarto option is not correctly picked up by RStudio IDE. Could you open an issue there maybe to report ? 

Thank you !

**Comment by cderv** [comment]
> By default when running code chunks it uses the project as the working directory and when rendering it uses the folder where the file is saved.

About this, is it in RStudio IDE when running code interactively ? I am surprised by this behavior and not sure I can reproduce. Though maybe this has to do with RStudio IDE version and you should check latest version (or even daily version to try out). 

For me in a quarto project, running chunk interactively in a file inside a sub folder of a quarto project will use the file folder as working directory for R.

**Reply by mcanouil** [reply]
> For me in a quarto project, running chunk interactively in a file inside a sub folder of a quarto project will use the file folder as working directory for R.

This would mean that the current session changes the working directory as you switch between files.
This would be really a hard to guess behaviour and could lead to real debugging nightmare.

**Reply by victorfeagins** [reply]
I had the behavior you described most of the semester then I tweaked some rstudio options and since then I get the behavior of the interactive chunks using the project as the working directory instead of the file location.

**Reply by victorfeagins** [reply]
Version Info for my Rstudio:
RStudio 2023.06.1+524 "Mountain Hydrangea" Release (547dcf861cac0253a8abb52c135e44e02ba407a1, 2023-07-07) for windows
Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) RStudio/2023.06.1+524 Chrome/110.0.5481.208 Electron/23.3.0 Safari/537.36

**Reply by mcanouil** [reply]
There is indeed an option in RStudio to define the working directory at least for RMarkdown document. I suggest you look for that as there are not that many parameters.

**Reply by cderv** [reply]
> This would mean that the current session changes the working directory as you switch between files.
This would be really a hard to guess behaviour and could lead to real debugging nightmare.

I am not sure I understand your point here. The R console itself does not change the working directory based on open file in source pane. If you run R code directly in the console it does not change. 
It adapts it for chunks in case of interactive rendering so that the code executed gives the same results as during rendering. 

RStudio IDE will respect the option `execute-dir`. If changed to
````yaml
project:
  execute-dir: project
````
then `quarto render` and interactive rendering are giving the same results and using project as root. 

> There is indeed an option in RStudio to define the working directory at least for RMarkdown document. 

I don't think there is for Quarto. Only for R Markdown. For Quarto, it works through the `_quarto.yml` option. 

> Version Info for my Rstudio:
> RStudio 2023.06.1+524 "Mountain Hydrangea" Release (547dcf861cac0253a8abb52c135e44e02ba407a1, 2023-07-07) for windows
> Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) RStudio/2023.06.1+524 Chrome/110.0.5481.208 Electron/23.3.0 Safari/537.36


@victorfeagins Support for this was added in 
- https://github.com/rstudio/rstudio/pull/11370
which landed first in `2023.12.1+402` so you need to update your IDE. 

It was released in January: https://docs.posit.co/ide/news/#rstudio-2023.12.1
